File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
###DFS
answer=0
def dfs(idx,val,numbers,target):
    global answer
    #탈출조건
    if idx==len(numbers):
        if val==target:
            answer+=1
        return
    #수행동작
    dfs(idx+1,val+numbers[idx],numbers,target)
    dfs(idx+1,val-numbers[idx],numbers,target)
def solution(numbers, target):
    dfs(0,0,numbers,target)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java

class Solution {
    static int ansCnt = 0;
    public int solution(int[] numbers, int target) {
        
        int[] arr = new int[numbers.length];
        
        for (int i = 0; i < arr.length; i++) {
            arr[i] = numbers[i];
        }
        dfs(1, arr[0], arr.length, target, arr);
        dfs(1, -arr[0], arr.length, target, arr);
        int answer = ansCnt;
        return answer;
    }
    public static void dfs(int depth, int total, int endPoint, int target, int[] arr) {
        if (depth == endPoint) {
            if (total == target) {
                ansCnt += 1;
            }
            return;
        }
        dfs(depth + 1, total + arr[depth], endPoint, target, arr);
        dfs(depth + 1, total - arr[depth], endPoint, target, arr);
    }
}